stan/math/opencl/matrix_cl.hpp

```diff
@@ -308,10 +306,9 @@ class matrix_cl<T, enable_if_arithmetic<T>> {
    * @throw <code>std::system_error</code> if the
    * matrices do not have matching dimensions
    */
-  explicit matrix_cl(
-      const Eigen::Ref<const Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>>&
-          A,
-      matrix_cl_view partial_view = matrix_cl_view::Entire)
+  template <typename Mat, require_eigen_t<std::is_arithmetic, Mat>...>
+  explicit matrix_cl(Mat&& A,
```

Comment:
I want to preserve l/r value-ness along with const so for the bug we have right now we can use `is_r_value` to make a local copy when we move the data to the device. If the param coming in is not const we can give OpenCL the host pointer so on CPUs we don't have to make a copy

Reply:
> I want to preserve l/r value-ness along with const

Of what? I don't understand what you want to say here. This is not forwarding `A` to any other function. I don't think this change is a solution for that bug.

Reply:
>> I want to preserve l/r value-ness along with const

> Of what? I don't understand what you want to say here. This is not forwarding A to any other function. 

It's not doing anything right now, but for zero copy you can't give OpenCL a const object's memory (because then that objects memory is gone violating constness). Having a generic template is going to let me in another PR look at both the valueness of the object (`A`) as well as it's constness so we can do things like

```
if (is_rvalue<Mat>::value) {
  // do a thing to fix the bug
} 
if (!is_const<Mat>::value) {
 // do the zero copy stuff
}

> I don't think this change is a solution for that bug.

Yeah not that I understand that bug more I agree. Though it lets us get to the solution

Reply:
Now I understand where you are going with this. Again if we want to support `Eigen::Map` or `Eigen::Block` without extra copies we need special handling. No need to do it in this PR.